File: 30 days of code/__ Java/Day 12 - Inheritance.java
/**
 *              INHERITANCE | JAVA
 * Domain       30 DAYS OF CODE
 * Sub domain   DAY 12
 */

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import java.util.stream.IntStream;

public class Main {
    public static void main(String[] args) throws FileNotFoundException {
        File file = new File("input.txt");
        Scanner scan = new Scanner(file);
        {
            String firstName = scan.next();
            String lastName = scan.next();
            int id = scan.nextInt();
            int[] testScores = new int[scan.nextInt()];
            for (int i = 0; i < testScores.length; ++i) {
                testScores[i] = scan.nextInt();
            }
            Student student = new Student(firstName, lastName, id, testScores);
            student.printPerson();
            System.out.println("Grade: " + student.calculate());
        }
        scan.close();
    }
}

class Person {
    private String firstName, lastName;
    private int id;

    Person(String firstName, String lastName, int id) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.id = id;
    }

    void printPerson() {
        System.out.println("Name: " + lastName + ", " + firstName + "\nID: " + id);
    }
}

class Student extends Person {
    private int[] testScores;

    Student(String firstName, String lastName, int id, int[] testScores) {
        super(firstName, lastName, id);
        this.testScores = testScores;
    }

    char calculate() {
        int total = IntStream.of(testScores).sum();// 0;
        int mean = total / testScores.length;
        if (mean >= 90 && mean <= 100) {
            return 'O';
        } else if (mean >= 80) {
            return 'E';
        } else if (mean >= 70) {
            return 'A';
        } else if (mean >= 55) {
            return 'P';
        } else if (mean >= 40) {
            return 'D';
        } else {
            return 'T';
        }
    }
}

File: 30 days of code/__ C#/Day 12 - Inheritance.cs
/**
 *              INHERITANCE | C#
 * Domain       30 DAYS OF CODE
 * Sub domain   DAY 12
 */

using System;
using System.IO;
using System.Linq;

class Program
{
    static void Main(string[] args)
    {
        var reader = new InputReader(false);
        var line = reader.NextArrString;
        var testScores = reader.NextArrInt;
        reader.Close();

        Student student = new Student(line[0], line[1], int.Parse(line[2]), testScores);
        student.Print();
        Console.WriteLine("Grade: " + student.Calculate());
    }
}

class Person
{
    private readonly string firstName;
    private readonly string lastName;
    private readonly int id;

    public Person(string firstName, string lastName, int id)
    {
        this.firstName = firstName;
        this.lastName = lastName;
        this.id = id;
    }

    public void Print() => Console.WriteLine($"Name: {lastName} {firstName}\nID: {id}";
}

class Student : Person
{
    private int[] testScores;

    public Student(string firstName, string lastName, int id, int[] testScores)
        : base(firstName, lastName, id)
    {
        this.testScores = testScores;
    }

    public char Calculate()
    {
        int total = testScores.Sum();
        int mean = total / testScores.Length;

        switch (mean)
        {
            case int i when i >= 90 && i <= 100:
                return 'O';
            case int i when i >= 80:
                return 'E';
            case int i when i >= 70:
                return 'A';
            case int i when i >= 55:
                return 'P';
            case int i when i >= 40:
                return 'D';
            default:
                return 'T';
        }
    }
}

class InputReader
{
    private TextReader input;

    public InputReader(bool fromFile)
    {
        input = fromFile ? new StreamReader("input.txt") : Console.In;
    }

    public string NextLine => input.ReadLine();

    public int[] NextArrInt => Array.ConvertAll(NextLine.Split(), int.Parse);

    public string[] NextArrString => NextLine.Split();

    public void Close() => input.Close();
}
